src/vmm/src/vstate/vm.rs

```diff
@@ -276,14 +499,215 @@ impl Vm {
         file.sync_all()
             .map_err(|err| MemoryBackingFile("sync_all", err))
     }
+
+    /// Register a device IRQ
+    pub fn register_irq(&self, fd: &EventFd, gsi: u32) -> Result<(), errno::Error> {
+        self.common.fd.register_irqfd(fd, gsi)?;
+
+        let mut entry = kvm_irq_routing_entry {
+            gsi,
+            type_: KVM_IRQ_ROUTING_IRQCHIP,
+            ..Default::default()
+        };
+        #[cfg(target_arch = "x86_64")]
+        {
+            entry.u.irqchip.irqchip = KVM_IRQCHIP_IOAPIC;
+        }
+        #[cfg(target_arch = "aarch64")]
+        {
+            entry.u.irqchip.irqchip = 0;
+        }
+        entry.u.irqchip.pin = gsi;
+
+        self.common
+            .interrupts
+            .lock()
+            .expect("Poisoned lock")
+            .insert(
+                gsi,
+                RoutingEntry {
+                    entry,
+                    masked: false,
+                },
+            );
+        Ok(())
+    }
+
+    /// Register an MSI device interrupt
+    pub fn register_msi(
+        &self,
+        route: &MsiVector,
+        masked: bool,
+        config: MsiIrqSourceConfig,
+    ) -> Result<(), errno::Error> {
+        let mut entry = kvm_irq_routing_entry {
+            gsi: route.gsi,
+            type_: KVM_IRQ_ROUTING_MSI,
+            ..Default::default()
+        };
+        entry.u.msi.address_lo = config.low_addr;
+        entry.u.msi.address_hi = config.high_addr;
+        entry.u.msi.data = config.data;
+
+        if self.common.fd.check_extension(kvm_ioctls::Cap::MsiDevid) {
+            // On AArch64, there is limitation on the range of the 'devid',
+            // it cannot be greater than 65536 (the max of u16).
```

Comment:
Can you provide a link to the source?

Reply:
Thanks for spotting this. I looked again into the Linux code and couldn't find any limitation specific to Aarch64. The KVM documentation does says: " For PCI, this is usually a BDF identifier in the lower 16 bits.". in our case these are the only bits that are non zero so we can just pass `config.devid`.